### ISSUE TITLE
The translate Behavior not working when I18n::setLocale('en@currency=EUR') is set

### ISSUE THREAD

**Issue by infoliv** [cakephp/cakephp#16062]
This is a (multiple allowed):

* [ ] bug
* [x] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 4.2.9
* Platform and Target: Apache2, PHP 7.4.24

### What you did
I use I18n where my routes use languages with 2 characters ("fr", "en", "it", ...)
```
// Route For example 
$builder
        ->connect(
            '/{lang}/{controller}/{action}/*', 
            []
        )
        ->setPatterns([
            'lang' => 'fr|en|it'
        ])
        ->setPersist(['lang']);
```
As the country is not defined, when I do
```
echo __(
   'Hi {0}, your balance on the {1,date} is {2,number,currency}',
    ['Charles', new FrozenTime('2014-01-13 11:12:00'), 1354.37]
);
```
the currency appears with the symbol `¤`

### What happened
I've tried to force the currency to `€` in the `AppController::initialize()` like that :
```
if ($this->request->getParam('lang')) {
            I18n::setLocale($this->request->getParam('lang') . '@currency=EUR');
        }
```
The currency is now displayed with `€` in my previous example but the Translate Behavior not working anymore.

### What you expected to happen
Would it be possible to enhance the Translate Behavior so that it could filter `@currency=EUR` ?

https://stackoverflow.com/questions/69679718/is-it-correct-i18nsetlocaleen-fr-in-cakephp-4/69681138?noredirect=1#comment123178945_69681138

**Comment by ADmad** [comment]
The locale string is used to load the `po` files too, so you will have problems with translating static strings too.

It would be better if you map your 2 chars language in URL to a locale with country name (e.g. `en_UK`) and use that for `I18n::setLocale()`. For e.g. my [I18n](https://github.com/admad/cakephp-i18n#i18nmiddleware) allows doing that.

**Comment by ndm2** [comment]
> The locale string is used to load the `po` files too, so you will have problems with translating static strings too.

The message file loader uses `\Locale::parseLocale()` and then grabs only the language and region parts, so that should work fine I think? 🤔 

https://github.com/cakephp/cakephp/blob/e230474b7b09cde40b0d8fe5923381878d5bc971/src/I18n/MessagesFileLoader.php#L153-L158

**Comment by markstory** [comment]
Marked as an enhancement but this is also in the realm of bug too.

**Comment by othercorey** [comment]
> Marked as an enhancement but this is also in the realm of bug too.

What part should be considered a bug and fixed?

**Comment by ndm2** [comment]
I'd say the fact that the translate behavior uses the raw locale string (eg `en_US@currency=EUR`) instead of just the language and region parts like the message file loader does (eg `en_US`).

Not sure though if there are maybe situations where one actually wants/needs to use the possible additional information from the raw string...

**Comment by ndm2** [comment]
Might still be nice to have. 5.x could be an opportunity for such a possible BC break

**Comment by othercorey** [comment]
Do you have a proposal?

**Comment by ndm2** [comment]
No, I only remember trying it some time back and running into some problems. Maybe it's still shelved somewhere, I'll look for it 🙃

**Comment by ndm2** [comment]
OK I don't have it anymore, but I can give it another try.

I'm wondering, does anyone know whether it would make sense to include the script additionally to language and region? For example `zh_Hans_HK` (Simplified) vs `zh_Hant_HK` (Traditional), or `sr_Cyrl_BA` (Cyrillic) vs `sr_Latn_BA` (Latin).

**Comment by jorisvaesen** [comment]
Had the same issue couple of weeks ago and fixed it with mapping the URL lang to the full locale as @ADmad suggested. What would hold back allowing only full locales (lang_REGION) on I18n::setLocale as of 5.0?

**Comment by ADmad** [comment]
> What would hold back allowing only full locales (lang_REGION) on I18n::setLocale as of 5.0?

This issue is regarding the Translate behavior, not I18n::setLocale().
